unittests/unit/test_IKInitializer.cpp

```diff
@@ -0,0 +1,1094 @@
+          std::vector<Eigen::Vector3s> markerTetherPoints;
+          markerTetherPoints.push_back(markerWorld);
+          markerTetherPoints.push_back(jointCenterEstimated);
+          server.createLine(
+              pair.first + ":" + markerName, markerTetherPoints, color);
+        }
+      }
+    }
+
+    server.saveFrame();
+  }
+  avgJointAngleError /= recoveredPoses.size();
+
+  if (numJointsCenterEstimatesCounted > 0)
+  {
+    avgJointCenterEstimateError /= numJointsCenterEstimatesCounted;
+  }
+  std::cout << "Marker error: " << markerError
+            << "m, joint center estimate error: " << avgJointCenterEstimateError
+            << "m" << std::endl;
+  std::cout << "Joint angle error: " << avgJointAngleError << std::endl;
+
+  if (saveToGUI)
+  {
+    server.writeFramesJson("../../../javascript/src/data/movement2.bin");
+  }
+
+  if (jointFailure)
+  {
+    return false;
+  }
+  if (avgJointCenterEstimateError > 0.05)
+  {
+    std::cout << "Joint error is too high on synthetic data!" << std::endl;
+    return false;
+  }
+  return true;
+}
+
+bool verifyJointCenterReconstructionOnSyntheticRandomPosesOsim(
+    std::string openSimPath)
+{
+  auto osim = OpenSimParser::parseOsim(openSimPath);
+  s_t targetHeight = 1.8;
+  s_t currentHeight = osim.skeleton->getHeight(
+      Eigen::VectorXs::Zero(osim.skeleton->getNumDofs()));
+  osim.skeleton->setBodyScales(
+      Eigen::VectorXs::Ones(osim.skeleton->getNumBodyNodes() * 3)
+      * (targetHeight / currentHeight));
+  Eigen::VectorXs goldGroupScales = osim.skeleton->getGroupScales();
+
+  std::vector<std::map<std::string, Eigen::Vector3s>> markerObservations;
+  std::vector<Eigen::VectorXs> poses;
+  std::vector<Eigen::VectorXs> jointCenters;
+  for (int t = 0; t < 50; t++)
+  {
+    Eigen::VectorXs pose = osim.skeleton->getRandomPose();
+    poses.push_back(pose);
+    osim.skeleton->setPositions(pose);
+    markerObservations.push_back(
+        osim.skeleton->getMarkerMapWorldPositions(osim.markersMap));
+    jointCenters.push_back(
+        osim.skeleton->getJointWorldPositions(osim.skeleton->getJoints()));
+  }
+
+  // Reset to neutral scale
+  osim.skeleton->setBodyScales(
+      Eigen::VectorXs::Ones(osim.skeleton->getNumBodyNodes() * 3));
+
+  IKInitializer initializer(
+      osim.skeleton, osim.markersMap, markerObservations, targetHeight);
+
+  // Initial guesses using MDS to center our subsequent least-squares on
+  initializer.closedFormMDSJointCenterSolver();
+  s_t pivotError = initializer.closedFormPivotFindingJointCenterSolver(true);
+  std::cout << "Pivot error avg: " << pivotError << "m" << std::endl;
+  initializer.recenterAxisJointsBasedOnBoneAngles(true);
+
+  s_t avgJointCenterEstimateError = 0.0;
+  int numJointsCenterEstimatesCounted = 0;
+  std::map<std::string, s_t> avgJointCenterEstimateErrorByJoint;
+  for (int t = 0; t < markerObservations.size(); t++)
+  {
+    auto jointMap = initializer.getVisibleJointCenters(t);
```

Comment:
What does it mean for joint center to be "visible"?

Reply:
I've renamed this to `getJointsAttachedToObservedMarkers`, which hopefully makes it more clear. It's the list of joints adjacent to markers that are visible on this timestep.